### PR TITLE
`get-consumer-group-lag` tool

### LINKED ISSUE #488: Consumer-group analysis tooling
The MCP server can produce to topics, consume from them, and (with #480) read partition watermarks — but it cannot tell a caller what consumer groups exist, who's in one, what they're consuming, or how far behind they are.
Common debugging workflows fall off the cliff:

- "What consumer groups are running on this cluster?"
- "Is group X actually consuming or stuck? Who are its members? What's it assigned to?"
- "What's group X's lag on topic Y right now — the live snapshot, not a metrics time-series?"

The metrics path covers lag as a time-series via the `consumer_lag_offsets` metric (see `docs/workflows/metrics-example-workflows.md`). That answers "plot lag over time for a group I already know about" — not "what groups exist on this cluster?" or "who's in this group right now?" The two views are complementary, not substitutes.

## Approach

Three read-only Kafka tools, all wrapping `KafkaJS.Admin` natively — same client path the rest of the Kafka domain takes (`consume-kafka-messages`, `list-topics`, `create-topics`, and the new `get-partition-offsets` from #480 all go native).

The REST proxy (`/kafka/v3/clusters/{id}/consumer-groups/*`) is available and pre-generated in `src/confluent/openapi-schema.d.ts`, but adopting it for these tools would introduce a second admin surface alongside the native path for no real win. The only thing REST adds is server-side lag arithmetic, which is one `BigInt` subtraction client-side. Native also gives us `listGroups` state filtering that the REST v3 surface doesn't expose.

All three tools share the same wiring decisions:

- Predicate: `kafkaBootstrapOrOAuth` (same as `consume-messages`, `produce-message`, `get-partition-offsets`).
- Category: `ToolCategory.Kafka`.
- Annotation: `READ_ONLY`.
- Standard `cluster_id` / `environment_id` optional args for OAuth connections.
- Response payloads surfaced via `createStructuredResponse` (the `structuredContent` channel #480 introduces).

### LINKED ISSUE #491: `get-consumer-group-lag` tool
Third and heaviest of the consumer-group analysis tools.
Combines `admin.fetchOffsets({groupId})` (committed offsets) with `admin.fetchTopicOffsets(topic)` (high watermarks) and emits per-(topic, partition) `{committed, high, lag}` rows.

Must reuse the watermark helper extracted by #480 — do not re-inline `admin.fetchTopicOffsets` at a second call site.
Whichever PR ships first introduces the helper module under `src/confluent/tools/handlers/kafka/` (per #480's coordination clause); this tool imports it.

## Tool surface

- Name: `get-consumer-group-lag`.
- Predicate: `kafkaBootstrapOrOAuth`.
- Category: `ToolCategory.Kafka`.
- Annotation: `READ_ONLY`.

## Input schema

```typescript
z.object({
  groupId: z
    .string()
    .min(1)
    .describe(
      "Consumer group ID to compute lag for. Discover groups via list-consumer-groups.",
    ),
  topics: z
    .array(z.string().min(1))
    .nonempty()
    .optional()
    .describe(
      "Restrict lag computation to these topics. " +
        "Omit to compute lag for every topic the group has committed offsets on.",
    ),
  cluster_id: z
    .string()
    .optional()
    .describe(
      "Confluent Cloud logical Kafka cluster ID (lkc-...). Discover via list-clusters.",
    ),
  environment_id: z
    .string()
    .optional()
    .describe(
      "Confluent Cloud environment ID (env-...) that owns the cluster. Discover via list-environments.",
    ),
});
```

`.nonempty()` on `topics` rejects `[]` — an empty filter is ambiguous (match nothing or match everything?), same call as `list-consumer-groups`' filter arrays.

## Response shape

```typescript
type GetConsumerGroupLagResponse = {
  groupId: string;
  topics: Array<{
    topic: string;
    partitions: Array<{
      partition: number;
      committedOffset: string | null; // int64 as digit-only string; null when never committed
      highWatermark: string; // int64 as digit-only string
      lag: number | null; // BigInt(high) - BigInt(committed), narrowed; null when never committed
      metadata: string | null;
      leaderEpoch: number | null;
    }>;
  }>;
  totalLag: number;
};
```

Same int64-as-string convention `get-partition-offsets` and `consume-messages.start.offset` established.
`lag` is a Number for caller ergonomics, narrowed from the BigInt diff via the same `bigIntToSafeMessageCount`-style helper #480 introduces.

Text summary: `Consumer group "<id>" has <totalLag> message(s) of lag across <N> topic(s).`

`totalLag` sums every non-null `lag` value across every partition.
Null-lag partitions (never committed) are excluded from the sum but still appear in the per-partition rows.

## Lag computation

```typescript
const diff = BigInt(highWatermark) - BigInt(committedOffset);
const lag = bigIntToSafeMessageCount(diff, { groupId, topic, partition });
```

The helper's overflow guard (saturate at `Number.MAX_SAFE_INTEGER` with a `Wacky --` log) carries forward.
Realistic magnitudes will not overflow; the guard is there because #480 already wrote it.

### When `committed > high`

Can happen transiently during a rebalance, or when the consumer's committed offset reflects reads at a different isolation level than the watermark fetch.
Surface the resulting negative `lag` as-is; do not clamp.
A negative lag is diagnostic — clamping to zero hides a real (if rare) state.
Document in the field's `.describe()` text that `lag` may briefly be negative during a rebalance.

This is _not_ a `Wacky --` case — it's a known race, not an impossible-codepath situation.
No log.

### When the group has no committed offset for a partition

`fetchOffsets` returns `offset: "-1"` for partitions the group has never committed to.
Treat these specially:

- `committedOffset: null`
- `lag: null`
- `metadata` and `leaderEpoch` as the broker returned them

Don't synthesize lag against `-1` — that would produce a misleadingly enormous positive number.

## Topic filtering

When `topics` is supplied:

- Pass it through to `fetchOffsets({groupId, topics})` — the admin call filters server-side.
- Validate that each requested topic exists on the cluster (via `admin.fetchTopicMetadata` — same call `get-partition-offsets` makes).
  A topic the group has never committed to is not an error; an unknown topic is.

When `topics` is omitted:

- `fetchOffsets({groupId})` returns offsets for every topic the group has touched.
- For each unique topic in the response, call the shared watermark helper.

Memoize the watermark-helper calls per topic across the request — a three-topic group with many partitions should make three admin round-trips, not one per partition.
The memoization pattern from `createWatermarkCache` in `consume-kafka-messages-handler.ts:529-542` is the model.

## Edge cases worth pinning in tests

- Group with committed offsets on N topics.
  Happy path; lag computed per (topic, partition); `totalLag` is the sum.
- Group that has never committed (zero topics in `fetchOffsets` response).
  `topics: []`, `totalLag: 0`, no error.
- Partition with `committedOffset === "-1"` (never committed).
  Surfaced as `committedOffset: null, lag: null`; not included in `totalLag`.
- Mix of committed and never-committed partitions within the same topic.
  Each partition row reflects its own state independently.
- `committed > high` (rebalance race).
  Negative `lag` surfaced as-is, no `Wacky --` log, partition included in `totalLag` with its negative contribution.
- `topics` filter narrows correctly.
  Only the requested topics in the response.
- `topics` filter includes a topic that doesn't exist on the cluster.
  Clean error: `Topic "<name>" not found on this cluster.`
- `topics` filter includes a topic the group has never committed to.
  Topic appears in the response with `partitions: []`, not an error.
- `topics: []` (empty array).
  Zod-level rejection via `.nonempty()`.
- Group ID that doesn't exist.
  Translate underlying error to `Consumer group "<id>" not found on this cluster.`
- Watermark helper module is the one shared with `get-partition-offsets`.
  Reviewer rejects the PR if a second `admin.fetchTopicOffsets` call site lands in the codebase.

## Acceptance criteria

- Handler under `src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts`.
- Registered in all four wiring points.
- Imports the shared watermark helper module — reject the PR if it re-inlines `admin.fetchTopicOffsets`.
- Unit tests pin every edge case above.
- Integration test against a real Confluent Cloud cluster, with a fixture that produces a known committed offset before the test so lag is deterministic.
- CHANGELOG entry.
- README mention.

## Relationships

Depends on the watermark helper from #480 (or extracts it here per the coordination clause — same story #479 has with #480).
Soft-depends on `list-consumer-groups` and `describe-consumer-group` for integration-test setup.

### PR DESCRIPTION
**Pull request by jlrobins**
## Summary of Changes

- New tool to fetch a consumer group's committed offsets vs the produced watermarks of the consumed topic(s) to then calculate the consumer lag.
- Had to factor out code handling fetching topic metadata to a new module / (function `normalizeFetchTopicMetadataResponse`)  shared between this handler and the consume messages handler due to same need to work around typescript issues present in the current release of the native Kafka client js binding. That issue has been fixed against main of confluent-kafka-javascript, but we need to wait for a new release. 
- Found another typescript issue (null vs undefined). Worked around it. Filed and raised a fix over at https://github.com/confluentinc/confluent-kafka-javascript/pull/490

Closes #491
Closes #488, is the final child.

### Manual testing instructions

1. Start up one or more consumers, then ask the agent to "describe the consumer groups" in the cluster, then "determine the lag for consumer group X":

...
"Check the lag again please"

> Called confluent (ctrl+o to expand)

> ⏺ spotify-consumer lag now:

| Partition | Committed | High watermark | Lag    | Δ vs. before           |
| --------- | --------- | -------------- | ------ | ---------------------- |
| 2         | 1         | 14,099         | 14,098 | 0 (stuck)              |
| 5         | 14        | 13,058         | 13,044 | committed +11, lag −10 |

> Total lag: 27,142 (down from 27,152 — net −10).

> Partition 5 is creeping forward; partition 2 hasn't budged. Partitions 0, 1, 3, 4 still uncommitted. The consumer's making very slow progress on one of six assigned partitions.

### Optional: Any additional details or context that should be provided?

- Centralized more unit and integration test fixtures.

## Pull request checklist

Please check if your PR fulfills the following (if applicable):

#### Tests

- [x] Added new
- [x] Updated existing
- [ ] Deleted existing

#### Release notes

- [x] Does anything in this PR need to be mentioned in the user-facing [CHANGELOG](https://github.com/confluentinc/mcp-confluent/blob/main/CHANGELOG.md)?

### CODE REVIEW

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+      }
+
+      // 2. For each topic+partition the group has committed to, fetch the
+      // partition's high watermark and compute lag = high - committed.
+
+      const watermarkCache = createWatermarkCache(admin);
+      const responseTopics: ConsumerGroupLagTopic[] = [];
+      let totalLag = 0;
+
+      for (const {
```
[the remark this answers is not in the document]
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts]: Fixed.

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+    if (isGroupIdNotFoundError(err)) return false;
+    throw err;
+  }
+  const desc = result.groups[0];
+  if (desc === undefined) return false;
+  if (desc.error !== undefined) {
+    if (desc.error.code === KafkaJS.ErrorCodes.ERR_GROUP_ID_NOT_FOUND) {
+      return false;
+    }
+    throw new Error(desc.error.message);
```
[the remark this answers is not in the document]
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts]: Fixed.

File: src/confluent/tools/handlers/kafka/partition-watermarks.ts [hunk 3]
```diff
@@ -14,6 +15,38 @@ export interface PartitionWatermark {
...
+): number {
+  const maxSafe = BigInt(Number.MAX_SAFE_INTEGER);
+  if (diff > maxSafe) {
+    logger.warn(
+      context,
+      "Wacky -- messageCount BigInt subtraction exceeds Number.MAX_SAFE_INTEGER; saturating to MAX_SAFE_INTEGER",
+    );
+    return Number.MAX_SAFE_INTEGER;
+  }
+  return Number(diff);
```
[the remark this answers is not in the document]
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/partition-watermarks.ts:56]: Fixed and tested.

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+  }
+}
+
+/**
+ * Build one {@link ConsumerGroupLagPartition} row from a committed offset and
+ * a paired watermark. Splits the never-committed sentinel (`offset === "-1"`)
+ * from the real arithmetic path so the latter remains a clean BigInt
+ * subtraction. Also returns the raw BigInt diff (or `null` for the
+ * never-committed case) so the caller can accumulate `totalLag` in BigInt
+ * and narrow once at the end — see [PR #508 Copilot comment 1].
```
[the remark this answers is not in the document]
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts]: Fixed.

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+  totalLagBigInt: bigint;
+}> {
+  const topics: ConsumerGroupLagTopic[] = [];
+  let totalLagBigInt = 0n;
+
+  for (const { topic, partitions: committedPartitions } of committedByTopic) {
+    const watermarks = await watermarkCache(topic);
+    const watermarkByPartition = new Map<number, PartitionWatermark>(
+      watermarks.map((w) => [w.partition, w]),
+    );
```
[the remark this answers is not in the document]
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts:405]: Fixed, tested.

File: src/confluent/tools/handlers/kafka/partition-watermarks.test.ts [hunk 6]
```diff
@@ -1,11 +1,93 @@
...
+      expect(narrowMessageCount(1n, {})).toBe(1);
+      expect(narrowMessageCount(1_000_000n, {})).toBe(1_000_000);
+    });
+
+    it("should pass small negative BigInt diffs through unchanged", () => {
+      // The lag tool surfaces a rebalance-race condition (committed >
+      // high watermark) as a small negative lag rather than clamping to
+      // zero — clamping would hide a real, if rare, state. The asymmetric
+      // saturation guard intentionally does not interfere with negative
+      // diffs of realistic magnitude.
```
[the remark this answers is not in the document]
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/partition-watermarks.test.ts]: Fixed.

File: src/confluent/tools/handlers/kafka/partition-watermarks.test.ts [hunk 7]
```diff
@@ -1,11 +1,93 @@
...
+      expect(narrowMessageCount(MIN_SAFE_BIGINT, { topic: "edge" })).toBe(
+        Number.MIN_SAFE_INTEGER,
+      );
+
+      expect(warnSpy).not.toHaveBeenCalled();
+    });
+
+    it("should saturate to Number.MIN_SAFE_INTEGER and emit a Wacky log when the diff is below the safe-integer boundary", () => {
+      // Caught by Copilot on PR #508: the guard was asymmetric (positive
+      // overflow only) and a diff below -(2^53 - 1) would silently lose
```
[the remark this answers is not in the document]
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/partition-watermarks.test.ts]: Fixed.

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.test.ts [hunk 8]
```diff
@@ -0,0 +1,863 @@
...
+      // saturation guard the per-partition path uses.
+      //
+      // Each partition's `committedOffset: "0"`, `highWatermark:
+      // String(MAX_SAFE_INTEGER)` gives a per-partition lag of exactly
+      // `MAX_SAFE_INTEGER` (in-range). Two such partitions sum to
+      // 2 * MAX_SAFE_INTEGER as a BigInt — well past the safe boundary,
+      // so the cross-partition narrow saturates and the test pins
+      // `totalLag === MAX_SAFE_INTEGER`. The buggy code would have
+      // accumulated to 2 * MAX_SAFE_INTEGER as a JS Number and failed
+      // this assertion.
```
[the remark this answers is not in the document]
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.test.ts]: Fixed.

File: CHANGELOG.md [hunk 9]
```diff
@@ -16,6 +16,7 @@ All notable changes to this MCP server will be documented in this file.
...
+- **`get-consumer-group-lag` tool.** Read-only live offset-lag computation for a single Kafka consumer group, combining `admin.fetchOffsets({groupId})` (committed offsets) with the shared partition-watermark helper (high watermarks) and BigInt-subtracting per partition.
```
[the remark this answers is not in the document]
Comment (jlrobins) [reply CHANGELOG.md:19]: The neighboring ones need to be reduced. This one is better.

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+    if (seenTopics.has(filterTopic)) continue;
+    try {
+      await watermarkCache(filterTopic);
+    } catch (err) {
+      if (isUnknownTopicError(err)) {
+        return { kind: "topicNotFound", missingTopic: filterTopic };
+      }
+      throw err;
+    }
+    topics.push({ topic: filterTopic, partitions: [] });
```
[the remark this answers is not in the document]
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts]: Switched to use less expensive `fetchTopicMetadata()`.

File: src/confluent/tools/handlers/kafka/partition-watermarks.test.ts [hunk 11]
```diff
@@ -1,11 +1,94 @@
...
+      // distinguishes this branch from the saturate branch — the return
+      // value is identical at the boundary.
+      const warnSpy = vi.spyOn(logger, "warn").mockImplementation(() => {});
+
+      expect(narrowMessageCount(MAX_SAFE_BIGINT, { topic: "edge" })).toBe(
+        Number.MAX_SAFE_INTEGER,
+      );
+
+      expect(warnSpy).not.toHaveBeenCalled();
+    });
```
[the remark this answers is not in the document]
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/partition-watermarks.test.ts:44]: vitest.config.ts:15 has restoreMocks: true, and the unit-tests rule documents it explicitly. This suggestion is a wrong-positive.

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+        committedOffset: null,
+        highWatermark: watermark.high,
+        lag: null,
+        metadata: committed.metadata,
+        leaderEpoch: committed.leaderEpoch,
+      },
+      lagBigInt: null,
+    };
+  }
+  const diff = BigInt(watermark.high) - BigInt(committed.offset);
```
[the remark this answers is not in the document]
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts:312]: Handled.

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+      }
+      throw err;
+    }
+    const found = metadata.topics.some((t) => t.name === filterTopic);
+    if (!found) {
+      return { kind: "topicNotFound", missingTopic: filterTopic };
+    }
+    topics.push({ topic: filterTopic, partitions: [] });
+  }
+  return { kind: "resolved", topics };
```
[the remark this answers is not in the document]
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts]: Applied batching.

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts [hunk 14]
```diff
@@ -0,0 +1,476 @@
...
+ * `ERR__UNKNOWN_TOPIC` (local-side, double underscore — fires when the
+ * client's own metadata cache rejects the topic before the broker is
+ * consulted). Both funnel into the same friendly "topic not found"
+ * message; same posture get-partition-offsets takes.
+ */
+function isUnknownTopicError(err: unknown): boolean {
+  if (!(err instanceof KafkaJS.KafkaJSError)) return false;
+  return (
+    err.code === KafkaJS.ErrorCodes.ERR_UNKNOWN_TOPIC_OR_PART ||
+    err.code === KafkaJS.ErrorCodes.ERR__UNKNOWN_TOPIC
```
[the remark this answers is not in the document]
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts]: Broadened and tested.

File: src/confluent/tools/handlers/kafka/describe-consumer-group-handler.ts [hunk 15]
```diff
@@ -142,20 +150,23 @@ export class DescribeConsumerGroupHandler extends BaseToolHandler {
...
           groupDesc.error.code === KafkaJS.ErrorCodes.ERR_GROUP_ID_NOT_FOUND
         ) {
-          return this.createResponse(notFoundMessage(parsed.groupId), true);
+          return this.createResponse(
+            notFoundGroupMessage(parsed.groupId),
+            true,
+          );
         }
         return this.createResponse(groupDesc.error.message, true);
       }
```
[the remark this answers is not in the document]
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/describe-consumer-group-handler.ts]: Fixed.

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts [hunk 16]
```diff
@@ -0,0 +1,566 @@
...
+      if (!metadata.some((t) => t.name === topic)) {
+        return topic;
+      }
+    } catch (err) {
+      if (isUnknownTopicError(err)) return topic;
+      throw err;
+    }
+  }
+  return null;
+}
```
Comment (bcscc) [inline src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts]: Seems to be used once as a fallback when batched `fetchTopicMetadata` rejects with an unknown-topic code but doesn't name the missing topic. Would a more simiplier error like `Topic not found in [<list>]` also work and avoids extra logic?

File: src/confluent/tools/handlers/kafka/consumer-group-not-found.ts
Comment (bcscc) [inline src/confluent/tools/handlers/kafka/consumer-group-not-found.ts:1]: `consumer-group-not-found.ts` overlaps with existing `consumer-group-enums.ts`. Maybe merge `consumer-group-not-found.ts` into a broader `consumer-group-helpers.ts` (or into the enums file) for better organization?

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts [hunk 17]
```diff
@@ -0,0 +1,566 @@
...
+ * consumer group, returning per-(topic, partition) {committed, high, lag}
+ * rows and a total lag count. Combines `admin.fetchOffsets({groupId})`
+ * (committed offsets) with the shared `fetchPartitionWatermarks` helper
+ * (high watermarks), then BigInt-subtracts per partition. A partition the
+ * group has never committed to surfaces as {committedOffset: null,
+ * lag: null} and is excluded from the `totalLag` sum. Lag may briefly go
+ * negative during a consumer-group rebalance (committed offset captured
+ * at a different isolation level than the watermark snapshot); the value
+ * is surfaced as-is rather than clamped, since clamping would hide a real
+ * if rare state. Closes #491.
```
Comment (bcscc) [inline src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts]:
```suggestion
 * if rare state.
```

File: src/confluent/tools/handlers/kafka/partition-watermarks.test.ts
Comment (bcscc) [inline src/confluent/tools/handlers/kafka/partition-watermarks.test.ts]:
```suggestion
      // `get-partition-offsets` handler read it. Narrowing here
```

File: src/confluent/tools/handlers/kafka/consumer-group-not-found.ts [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+      KafkaJS.ErrorCodes.ERR_GROUP_ID_NOT_FOUND
+  );
+}
+
+/**
+ * Third path the broker uses to report an unknown group: a
+ * successful-shaped {@link GroupDescription} with no `.error` field, state
+ * `Dead`, no members, and empty `protocol` / `partitionAssignor` strings.
+ * Confluent Cloud's brokers were observed taking this path
+ * (integration-test discovery, May 2026) rather than the documented
```
Comment (bcscc) [inline src/confluent/tools/handlers/kafka/consumer-group-not-found.ts]:
```suggestion
 * Confluent Cloud's brokers were observed taking this path
 * rather than the documented
```

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.integration.test.ts [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+describe("get-consumer-group-lag-handler", { tags: [Tag.KAFKA] }, () => {
+  if (handler.enabledConnectionIds(runtime).length === 0) {
+    it.skip("requires kafka.bootstrap_servers config", () => {});
+    return;
+  }
+
+  let admin: KafkaJS.Admin;
+  let producer: KafkaJS.Producer;
+  const topic = uniqueName("lag");
+  // Second topic stays exist-but-untouched so test #4 below can prove the
```
Comment (bcscc) [inline src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.integration.test.ts]:
```suggestion
  // Second topic stays exist-but-untouched so below can prove the
```

File: src/confluent/tools/handlers/kafka/topic-metadata.ts
Comment (bcscc) [inline src/confluent/tools/handlers/kafka/topic-metadata.ts:1]: Should we have an open ticket to clean this up once the upstream fix is in?

File: src/confluent/tools/handlers/kafka/partition-watermarks.ts
Comment (bcscc) [inline src/confluent/tools/handlers/kafka/partition-watermarks.ts]:
```suggestion
 * {@link createWatermarkCache}) and standalone
```

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts [hunk 20]
```diff
@@ -0,0 +1,566 @@
...
+/**
+ * Best-effort group-existence probe used to disambiguate the
+ * `fetchOffsets`-resolves-empty path. Returns `true` if the group is known
+ * to exist on the cluster, `false` if the broker reported it as unknown
+ * via any of the three shapes the project handles (top-level rejection,
+ * per-group `GroupDescription.error` with the same code, or the
+ * dead-tombstone — see {@link isUnknownGroupTombstone} for why the
+ * tombstone path matters on Confluent Cloud). Propagates any other error.
+ */
+async function groupExists(
```
Comment (bcscc) [inline src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts]: Question, this seems to be an extension of `describe-consumer-group-handler`'s decoding logic from lines 132-174 where `describeGroups([id])` is decoded into one of: not-found, real-error, real-group. Would it make sense to lift into these into one function used by both handlers in `consumer-group-not-found.ts`?

**Review by bcscc — COMMENTED** [review]
Thanks, have some questions and comments

File: src/confluent/tools/handlers/kafka/topic-metadata.ts
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/topic-metadata.ts:1]: https://github.com/confluentinc/mcp-confluent/issues/514

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts (hunk 17 above)
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts]: Removed, and friends like it.

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts (hunk 16 above)
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts]: Revised this section of code / made new notFoundTopicMessage().

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts (hunk 20 above)
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.ts]: Merged into shared describeGroupOutcome.

File: src/confluent/tools/handlers/kafka/consumer-group-not-found.ts
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/consumer-group-not-found.ts:1]: Merged and now in consumer-group-helpers.ts.

File: src/confluent/tools/handlers/kafka/partition-watermarks.test.ts
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/partition-watermarks.test.ts]: Added.

File: src/confluent/tools/handlers/kafka/consumer-group-not-found.ts (hunk 18 above)
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/consumer-group-not-found.ts]: Revised.

File: src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.integration.test.ts (hunk 19 above)
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/get-consumer-group-lag-handler.integration.test.ts]: Removed

File: src/confluent/tools/handlers/kafka/partition-watermarks.ts
Comment (jlrobins) [reply src/confluent/tools/handlers/kafka/partition-watermarks.ts]: Added.

**Review by bcscc — APPROVED** [review]
LGTM, thanks!
